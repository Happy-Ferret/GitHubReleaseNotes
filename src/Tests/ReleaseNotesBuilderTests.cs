namespace ReleaseNotesCompiler.Tests
{
    using System;
    using System.Linq;
    using ApprovalTests;
    using NUnit.Framework;
    using Octokit;

    [TestFixture]
    public class ReleaseNotesBuilderTests
    {
        [Test]
        public void NoCommitsNoIssues()
        {
            AcceptTest(0);
        }

        [Test]
        public void NoCommitsSomeIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SomeCommitsNoIssues()
        {
            AcceptTest(5);
        }

        [Test]
        public void SomeCommitsSomeIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SingularCommitsNoIssues()
        {
            AcceptTest(1);
        }

        [Test]
        public void SingularCommitsSomeIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"), CreateIssue(2, "Feature"), CreateIssue(3, "Improvement"));
        }

        [Test]
        public void SingularCommitsSingularIssues()
        {
            AcceptTest(1, CreateIssue(1, "Bug"));
        }

        [Test]
        public void NoCommitsSingularIssues()
        {
            AcceptTest(0, CreateIssue(1, "Bug"));
        }

        [Test]
        public void SomeCommitsSingularIssues()
        {
            AcceptTest(5, CreateIssue(1, "Bug"));
        }

        static void AcceptTest(int commits, params Issue[] issues)
        {
            var fakeClient = new FakeGitHubClient();

            fakeClient.Milestones.Add(CreateMilestone("1.2.3"));

            fakeClient.NumberOfCommits = commits;

            foreach (var issue in issues)
            {
                fakeClient.Issues.Add(issue);
            }

            var builder = new ReleaseNotesBuilder(fakeClient, "TestUser", "FakeRepo", "1.2.3");

            var notes = builder.BuildReleaseNotes().Result;

            Approvals.Verify(notes);
        }

        static Milestone CreateMilestone(string version)
        {
            return new Milestone
                {
                    Title = version,
                    Url = new Uri("https://github.com/Particular/FakeRepo/issues?q=milestone%3A" + version)
                };
        }

        static Issue CreateIssue(int number, params string[] labels)
        {
            return new Issue
                {
                    Number = number,
                    Title = "Issue " + number,
                    HtmlUrl = new Uri("http://example.com/" + number),
                    Body = "Some issue",
                    Labels = labels.Select(x => new Label { Name = x }).ToArray(),
                };
        }
    }
}