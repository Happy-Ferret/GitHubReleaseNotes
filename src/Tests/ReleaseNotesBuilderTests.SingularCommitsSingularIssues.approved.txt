As part of this release we had [1 commit](https://github.com/TestUser/FakeRepo/commits/1.2.3) which resulted in [1 issue](https://github.com/FakeRepo/issues/issues?milestone=0&state=closed) being closed.


__Bug__
- [__#1__](http://example.com/1) Issue 1

## Where to get it
You can download this release from [nuget](https://www.nuget.org/profiles/nservicebus/)